src/PrivateRoute.tsx

```diff
@@ -0,0 +1,30 @@
+import React, { FC, useContext } from 'react';
+import { Route, Redirect } from 'react-router-dom';
+import { AuthContext } from './Auth';
+import { ContextProps } from './types';
+
+type Props = { exact: true; path: string; component: FC<Props> };
+
+const PrivateRoute: FC<Props> = ({
+  component: RouteComponent,
+  exact,
+  path,
+}) => {
+  const { currentUser } = useContext<ContextProps>(AuthContext);
+
+  return (
+    <Route
+      exact={exact}
+      path={path}
+      render={(routeProps) =>
+        currentUser ? (
+          <RouteComponent {...routeProps} />
```

Comment:
@horitaka ここで以下のエラーが出てるのですがどう書くのが正しいのでしょうか
```
Type '{ history: History<PoorMansUnknown>; location: Location<PoorMansUnknown>; match: match<any>; staticContext?: StaticContext | undefined; }' is missing the following properties from type 'Props': exact, path, component
```

Reply:
上記のエラーはVSCode上でRouteComponentの部分にでてくるものです。
さらに...routePropsの部分には以下のエラーも出ています
```
(parameter) routeProps: RouteComponentProps<any, StaticContext, History.PoorMansUnknown>
Prop spreading is forbiddeneslint(react/jsx-props-no-spreading)
```

Reply:
６行目を以下のようにすればエラーは解消しないでしょうか？
```
type Props = { exact: true; path: string; component: FC<object> };
```

component: FC<object> は、propsとして受け取るcomponentのことだと思いますので、componentのpropsの型は何らかのobjectになると思います。

Reply:
修正してみたのですがだめでした。
https://github.com/kotsukotsuG/firebase-react-auth-typescript/pull/1/commits/7f530366d065e6b228319ada1a9acf198ba5dab7


少し、そもそも論にもどって質問したいのですが、ここのroutePropsには具体的には何が入っているのでしょう？
ルート(親？)のPropsなのでPrivateRouteコンポーネントに渡されたPropsすなわち、以下(9行目のあたり)が渡ってきていると思えばよいのでしょうか？
```
{
  component: RouteComponent,
  exact,
  path,
}
```

そして、呼び出し元では下になっているので
```
<PrivateRoute exact path="/" component={Home} />
```

結局routePropsとは以下になるという理解で良いのでしょうか？
```
{
  component: Home
  exact: true
  path: "/"
}

Reply:
修正方法が間違っていた点、及び、説明不足だった点は申し訳ありません。

> 少し、そもそも論にもどって質問したいのですが、ここのroutePropsには具体的には何が入っているのでしょう？

routePropsにはreact routerから渡されたpropsが入ります。

> ルート(親？)のPropsなのでPrivateRouteコンポーネントに渡されたPropsすなわち、以下(9行目のあたり)が渡ってきていると思えばよいのでしょうか？

いいえ、それとは違います。
上記の通り、react routerから渡されたpropsになります。

具体的にはここに書いてあるものになります。
https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Route.js#L44

locationというのはpathとかqueryパラメーターが入るようです。
他に何が入るかは、もうちょっとAPIリファレンスなどを見てみないとわかりませんでした。

開発ツールで確認をしてみたのですが、今回のアプリケーションの場合だと以下のようなものになるようです。
![スクリーンショット 2020-05-19 22 23 19](https://user-images.githubusercontent.com/20832866/82331696-6d802600-9a1f-11ea-81e9-815e7d6f9367.png)




修正の方法は正しくは以下でした。申し訳ありません。

**Home.tsx**
```diff
import React, { FC } from 'react';
+ import { RouteComponentProps } from 'react-router-dom'

import app from './base';

// type Props = {};

const handleClick = (): void => {
  app.auth().signOut();
};
- const Home: FC<void> = () => {
+ const Home: FC<RouteComponentProps> = () => {
  return (
```

**PrivateRoute.tsx**

```diff
import { Route, Redirect } from 'react-router-dom';
import { AuthContext } from './Auth';
import { ContextProps } from './types';
+ import { RouteComponentProps } from 'react-router-dom'

- type Props = { exact: true; path: string; component: FC<object> };
+ type Props = { exact: true; path: string; component: FC<RouteComponentProps> };
```

修正の方法はこちらの記事などが参考になるかと思います。
https://qiita.com/taneba/items/b50078e1ac9d1971d413

英語のこちらの記事も参考になりそうでした。
https://www.pluralsight.com/guides/react-router-typescript

Reply:
ちなみに、react dev toolsはご存知でしょうか？

今回のようにpropsの値が良くわからないというときやデバッグをするときに使えるかと思います。
https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en

Reply:
ありがとうございます。
Home.tsxのところですがちょっと細かいですがおしえてください。
```suggestion
- const Home: FC<RouteComponentProps> = () => {
+const Home: FC<RouteComponentProps> = (props) => {
```
HomeにはpropsとしてRouteComponentPropsが渡されるので、ジェネリクスで型を指定した上でpropsという引数も指定してあげないといけないのかと思っていたのですが、
特になくてもエラーはでないようです。

特に使わない引数であれば(RouteComponentPropsはHomeコンポーネント内部ではつかわれていません)、たとえ上記のコンポーネントから渡ってきたとしても記載する必要はないのでしょうか？

Reply:
```suggestion
- const Home: FC<RouteComponentProps> = () => {
+ const Home: FC<RouteComponentProps> = (props) => {
```
細かいですがHome.tsxの上記の修正いただいた箇所。引数にpropsを指定する必要はないのでしょうか？

Reply:
react dev toolsのご案内ありがとうございます。
昔、導入してすっかり忘れてましたので復習しておきます。

Reply:
> 細かいですがHome.tsxの上記の修正いただいた箇所。引数にpropsを指定する必要はないのでしょうか？

Homeの中でpropsを使って何かをするというのでないのであればpropsは書いておかなくても大丈夫ですね。
typescriptのエラーも出ないかと思います。

Reply:
なるほど。そういうものなのですね。